File: typescript/src/state/state.test.ts
import { describe, expect, it } from "vitest";
import { ConcreteStateA } from "./concrete-state-a.ts";
import { Context } from "./context.ts";
import type { State } from "./state.ts";

describe("State", () => {
  it("states alternate on each request", () => {
    const context = new Context(new ConcreteStateA());
    expect(context.getStateName()).toBe("ConcreteStateA");
    context.request();
    expect(context.getStateName()).toBe("ConcreteStateB");
    context.request();
    expect(context.getStateName()).toBe("ConcreteStateA");
  });

  it("the state decides the transition", () => {
    const stuck: State = {
      handle: () => {
        /* stays where it is */
      },
      name: () => "Stuck",
    };
    expect(new Context(stuck).request()).toBe("request() handled by Stuck, now in Stuck");
  });

  it("a state instance can serve several contexts", () => {
    const shared = new ConcreteStateA();
    const one = new Context(shared);
    const two = new Context(shared);
    one.request();
    expect(one.getStateName()).toBe("ConcreteStateB");
    expect(two.getStateName()).toBe("ConcreteStateA");
  });
});

File: java/src/test/java/com/penapereira/patterns/state/StateTest.java
package com.penapereira.patterns.state;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.junit.jupiter.api.Test;

class StateTest {

    @Test
    void statesAlternateOnEachRequest() {
        Context context = new Context(new ConcreteStateA());
        assertEquals("ConcreteStateA", context.getStateName());
        context.request();
        assertEquals("ConcreteStateB", context.getStateName());
        context.request();
        assertEquals("ConcreteStateA", context.getStateName());
        context.request();
        assertEquals("ConcreteStateB", context.getStateName());
    }

    @Test
    void theStateDecidesTheTransition() {
        State stuck = new State() {
            @Override
            public void handle(Context context) {
                // stays where it is
            }

            @Override
            public String name() {
                return "Stuck";
            }
        };
        Context context = new Context(stuck);
        assertEquals("request() handled by Stuck, now in Stuck", context.request());
    }

    @Test
    void aStateInstanceCanServeSeveralContexts() {
        State shared = new ConcreteStateA();
        Context one = new Context(shared);
        Context two = new Context(shared);
        one.request();
        assertEquals("ConcreteStateB", one.getStateName());
        assertEquals("ConcreteStateA", two.getStateName());
    }
}
